### PR TITLE
Fix/jca test coverage increase

### PR DESCRIPTION
**Pull request by mustafaiman**
Arquillian tests with embedded Glassfish Server

### CODE REVIEW

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/ConnectionFactoryImplTest.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jca;
+
+import com.hazelcast.core.*;
```
Comment (mesutcelik) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/ConnectionFactoryImplTest.java]: Can you please update your IDE not to include star imports?

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastTransactionImplTest.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        mockEventListener = mock(ConnectionEventListener.class);
+        managedConnection = managedConnectionFactory.createManagedConnection(null, null);
+        managedConnection.addConnectionEventListener(mockEventListener);
+        transaction = new HazelcastTransactionImpl(managedConnectionFactory, (ManagedConnectionImpl) managedConnection);
+
+
+    }
+
+    @Test
+    @Ignore
```
Comment (mesutcelik) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastTransactionImplTest.java]: Why do we have this Ignore annotation?

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastTransactionImplTest.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    public void testTransactionCommitShouldNotFireRemovedEventListener() throws ResourceException {
+        managedConnection.removeConnectionEventListener(mockEventListener);
+        transaction.begin();
+        transaction.commit();
+        verifyZeroInteractions(mockEventListener);
+    }
+
+    @Test
+    public void testGetLocalTransactionShouldCreateANewOneIfNotExists() throws ResourceException {
+        managedConnection.getLocalTransaction();
```
Comment (mesutcelik) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastTransactionImplTest.java]: what are we testing here?

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastTransactionImplTest.java (hunk 2 above)
Comment (mustafaiman) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastTransactionImplTest.java]: This is because ManagedConnectionImpl class never call its listener's localTransactionStarted method because of hardcoded logic. I thought it should call and asked @gurbuzali about it. He was cautious about changing the logic. I left the test there because I think it should be implemented but existing code would not pass the test.

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/ConnectionFactoryImplTest.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    @Test
+    public void testEquality() {
+        assertTrue(connectionFactory.equals(connectionFactory));
+        assertFalse(connectionFactory.equals(null));
+        assertFalse(connectionFactory.equals(new Object()));
+
+    }
+
+    @Test
+    public void testIdEqualityShouldDo() {
```
Comment (pveentjer) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/ConnectionFactoryImplTest.java]: Is this not another equals test?

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastTransactionImplTest.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Created by mustafaiman on 25/06/15.
+ */
+public class HazelcastTransactionImplTest extends HazelcastTestSupport{
+
+    HazelcastInstance instance;
```
Comment (pveentjer) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastTransactionImplTest.java]: Please use private for fields unless there is a reason not to do so.

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastTransactionImplTest.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+import javax.annotation.Resource;
+import javax.resource.ResourceException;
+import javax.resource.spi.ConnectionEvent;
+import javax.resource.spi.ConnectionEventListener;
+import javax.resource.spi.ManagedConnection;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
```
Comment (pveentjer) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastTransactionImplTest.java]: Please remove such comments. They don't provide any value.

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastTransactionImplTest.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        ResourceAdapterImpl resourceAdapter = new ResourceAdapterImpl();
+        resourceAdapter.setHazelcastInstance(instance);
+        ManagedConnectionFactoryImpl managedConnectionFactory = new ManagedConnectionFactoryImpl();
+        managedConnectionFactory.setResourceAdapter(resourceAdapter);
+        managedConnectionFactory.setConnectionTracingEvents(null);
+        mockEventListener = mock(ConnectionEventListener.class);
+        managedConnection = managedConnectionFactory.createManagedConnection(null, null);
+        managedConnection.addConnectionEventListener(mockEventListener);
+        transaction = new HazelcastTransactionImpl(managedConnectionFactory, (ManagedConnectionImpl) managedConnection);
+
```
Comment (pveentjer) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastTransactionImplTest.java]: Why these empty lines?

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastTransactionImplTest.java (hunk 3 above)
Comment (pveentjer) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastTransactionImplTest.java]: I agree. Should the returned value not be verified?

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/ManagedConnectionFactoryImplTest.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+import static org.junit.Assert.*;
+
+/**
+ * Created by mustafaiman on 25/06/15.
+ */
+@RunWith(HazelcastParallelClassRunner.class)
+@Category(QuickTest.class)
+public class ManagedConnectionFactoryImplTest extends HazelcastTestSupport {
+
+    ManagedConnectionFactory connectionFactory;
```
Comment (pveentjer) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/ManagedConnectionFactoryImplTest.java]: make private

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastTransactionImplTest.java (hunk 3 above)
Comment (mustafaiman) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastTransactionImplTest.java]: That was a mistakenly left there. Sorry

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastTransactionImplTest.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+import javax.resource.spi.ConnectionEventListener;
+import javax.resource.spi.ManagedConnection;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Created by mustafaiman on 25/06/15.
+ */
+public class HazelcastTransactionImplTest extends HazelcastTestSupport{
```
Comment (bilalyasar) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/HazelcastTransactionImplTest.java]: missing category

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/ManagedConnectionFactoryImplTest.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    public void testIdEqualityShouldDo() {
+        int id = 1234;
+        ManagedConnectionFactory c2 = new ManagedConnectionFactoryImpl();
+        ((ManagedConnectionFactoryImpl) connectionFactory).setId(id);
+        ((ManagedConnectionFactoryImpl) c2).setId(id);
+        assertTrue(c2.equals(connectionFactory));
+    }
+
+    @Test
+    public void testIdNotEqualityShouldReturnFalse() {
```
Comment (bilalyasar) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/ManagedConnectionFactoryImplTest.java]: what is the difference between this test and `testIdEqualityShouldDo` test.. seems they are same..

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/ManagedConnectionFactoryImplTest.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        Arrays.sort(returnItems);
+
+        assertArrayEquals(controlItems, returnItems);
+
+
+    }
+
+    @Test
+    public void testEquality() {
+        assertTrue(connectionFactory.equals(connectionFactory));
```
Comment (bilalyasar) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/ManagedConnectionFactoryImplTest.java]: what is the logic behind `connectionFactory.equals(connectionFactory)`?

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/ManagedConnectionFactoryImplTest.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+        assertFalse(connectionFactory.equals(null));
```
Comment (bilalyasar) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/ManagedConnectionFactoryImplTest.java]: `assertNotEquals` comes more readable to me. but it is just choice.

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/XATestWithJCA.java [hunk 13]
```diff
@@ -79,9 +82,76 @@ public void testPut() throws ResourceException {
...
+        multiMap.put(key,val1);
+        multiMap.put(key,val2);
+
+        connection.close();
+
+        MultiMap<String, String> m = instance.getMultiMap(name);
+        Collection<String> vals = m.get(key);
+        assertTrue(vals.contains(val1));
+        assertTrue(vals.contains(val2));
+        assertTrue(vals.size() == 2);
```
Comment (bilalyasar) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/XATestWithJCA.java]: here also i prefer `assertEquals`

Comment (pveentjer) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/XATestWithJCA.java]: +1. The advantage is that in case of failure, you know exactly what the values were. With assertTrue you loose that information unless you craft a message.

**Comment by mesutcelik** [comment]
@mustafaiman ,
are you finished with that?

**Comment by mustafaiman** [comment]
I will push integration tests to this same branch. I am working on it.

**Comment by mesutcelik** [comment]
Ok. next time you can work on smaller PRs. That is easy to digest for the reviewer and PR is not kept open for a long time... The aim is to merge PRs in shortest time possible...

**Comment by mesutcelik** [comment]
verify

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/ManagedConnectionFactoryImplTest.java [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+    @Test
+    public void testMatchExistingManagedConnection() throws ResourceException {
+        ResourceAdapterImpl mockResourceAdapter = mock(ResourceAdapterImpl.class);
+        when(mockResourceAdapter.getHazelcastInstance()).thenReturn(null);
+        ((ManagedConnectionFactoryImpl)connectionFactory).setResourceAdapter(mockResourceAdapter);
+
+        ((ManagedConnectionFactoryImpl)connectionFactory).setConnectionTracingEvents(null);
+
+        HashSet<ManagedConnection> managedConnections = new HashSet<ManagedConnection>();
+        for (int i=0; i < 9; i++) {
```
Comment (mesutcelik) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/ManagedConnectionFactoryImplTest.java]: what is the usage of this `for` loop?

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/XATestWithJCA.java [hunk 15]
```diff
@@ -79,6 +88,69 @@ public void testPut() throws ResourceException {
...
 
+    @Test
+    public void testTransactionalQueueShouldPollWhatWasOffered() throws ResourceException {
+        String name = randomString();
+        String item = randomString();
+
+        TransactionalQueue<String> queue = connection.getTransactionalQueue(name);
+        queue.offer(item);
+
+        connection.close();
```
Comment (mesutcelik) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/XATestWithJCA.java]: is there any specific reason not to close the connection within teardown method annotated with `@After`?

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/ManagedConnectionFactoryImplTest.java (hunk 14 above)
Comment (mustafaiman) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/ManagedConnectionFactoryImplTest.java]: mocks 10 more connections with different connection request info than the desired one. this ensures that ConnectionFactoryImpl.matchManagedConnections() can find a suitable connection without any security information limitation among other connections.

File: hazelcast-ra/hazelcast-jca/src/test/resources/ra.xml [hunk 16]
```diff
@@ -23,7 +23,7 @@
 
     <vendor-name>Hazelcast.com</vendor-name>
     <eis-type>Hazelcast</eis-type>
-    <resourceadapter-version>3.2-RC3-SNAPSHOT</resourceadapter-version>
+    <resourceadapter-version>3.6-SNAPSHOT</resourceadapter-version>
```
Comment (mesutcelik) [inline hazelcast-ra/hazelcast-jca/src/test/resources/ra.xml]: @bilalyasar ,
Are we updating this version per release script execution?

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/XATestWithJCA.java (hunk 15 above)
Comment (mustafaiman) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/XATestWithJCA.java]: data should persist in the instance after user connection is closed. close operation is a part of the test

**Comment by mesutcelik** [comment]
some minor comments...

@mustafaiman ,
that is excellent work! JCA had another meaning after this contribution :+1:

**Comment by mustafaiman** [comment]
thank you for the comment

File: hazelcast-ra/hazelcast-jca/src/test/resources/ra.xml (hunk 16 above)
Comment (bilalyasar) [inline hazelcast-ra/hazelcast-jca/src/test/resources/ra.xml]: yes, it is updated https://github.com/hazelcast/DevOps/blob/master/build-tool/change-version.sh#L16

File: hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/TestInContainer.java [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+
+        testBean.insertToSet(setname, key);
+        assertEquals(1, testBean.getSetSize(setname));
+        assertTrue(testBean.removeFromSet(setname, key));
+        assertEquals(0, testBean.getSetSize(setname));
+    }
+
+    @Test
+    @OperateOnDeployment("war-deployment")
+    @Ignore
```
Comment (gurbuzali) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/TestInContainer.java]: ignore ?

Comment (mesutcelik) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/TestInContainer.java]: He is gonna send it in another PR i guess. As this PR got bigger, we decided to split it up...

Comment (mustafaiman) [inline hazelcast-ra/hazelcast-jca/src/test/java/com/hazelcast/jca/TestInContainer.java]: yes that test had a problem. i will remove that ignore in another pr.
